services/images/local.go

```diff
@@ -161,7 +162,13 @@ func (l *local) Delete(ctx context.Context, req *imagesapi.DeleteImageRequest, _
 	log.G(ctx).WithField("name", req.Name).Debugf("delete image")
 
 	if err := l.store.Delete(ctx, req.Name); err != nil {
-		return nil, errdefs.ToGRPC(err)
+		if strings.HasPrefix(req.Name, "docker.io/library/") {
```

Comment:
I'm unsure about this change. This service doesn't do any caching. Does it have to worry about the prefix?

Reply:
The reason is related to 

- `nerdctl inspect`  shows no docker.io/library/ prefix.
- `crictl inspect`  shows docker.io/library/ prefix.
- `ctr i ls`  shows no docker.io/library/ prefix.

When the user removes an image thought `crictl`, it adds a prefix for it by default.
- So I have to double-check to delete the image without the prefix here as well.

If `docker.io/library/xxx:yyy` equals `xxx:yyy`, this logic change will have no side effect.

Reply:
I'm tempted to say: don't strip it (unless this is causing severe issues). Stripping the `docker.io/library` (and  `docker.io/`) prefixes mostly originates from remaining the UX on `docker`, which (when there was only 1 registry: the docker index (later "docker hub")), so images had no prefix. `ctr` probably strips the prefix to have a closer match with `docker` commands, but for most cases (and APIs), this should probably be considered part of the _presentation_, not what APIs returns, and make sure the API returns the canonical (full) image reference.

Adding to the above; the `docker.io` prefix as a whole is causing a lot of headaches (e.g., `docker.io` is not actually the domain of the registry (which currently is at `registry-1.docker.com`). We're actively exploring what options we have to get out of this confusing situation (but it's _complicated_ especially to remain backward compatible with existing container runtimes/tools). From that perspective, it'd be good to not introduce _more_ places where special handling is performed (unless there's a strong reason it's needed).

Reply:
/cc @ingshtrom (FYI)

(for others: @ingshtrom is one of my colleagues at docker exploring options around this 😅)

Reply:
- `nerdctl inspect` shows no `docker.io/library/` prefix.
-` crictl inspect` shows `docker.io/library/` prefix.
- `ctr i ls` shows no `docker.io/library/` prefix.

Another way to fix the issue mentioned is to make `crictl` not adding the prefix. I need to look into the code if this is doable.

Reply:
`nerdctl` is  designed to provide a UX that matches `docker`. It should probably be looked at _where_ that UX should be applied. Containerd in itself is (should be) less opinionated, which may mean: always assume canonical references. Some amount of compatibility is probably OK (which may mean convert _to_ (normalise to) a canonical reference on the way in, and leave it to products on top of containerd to decide on the UX they want to provide (which _may_ be "remove the prefix" when presenting).

We should really be careful not to leak docker's UX into containerd.

Reply:
I saved the original change in https://github.com/containerd/containerd/compare/main...pacoxu:image-clean-original?expand=1.

I reverted the `docker.io/library` trim in `services/images/local.go`. And I will look into `crictl` code to see if it can be optimized.